[Flight Recorder] Add more basic analysis to the script

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #133412

This is the first step to make sure we have a basic function of analyzer for FR in production.

- We want to use this script to find out abnormalities in collectives and report it to users.
- We also fixed some type errors.

- [Ongoing] Also we will add more unit tests to this script and make it modularized so that we can better maintain it.



cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @wz337 @wconstab @d4l3k @c-p-i-o

fduwjj committed:
[Flight Recorder] Add more basic analysis to the script

[ghstack-poisoned]

fduwjj committed:
Update on "[RFC + WIC][Flight Recorder] Add more basic analysis to the script"


This is the first step to make sure we have a basic function of analyzer for FR in production.
We want to use this script to find out abnormalities in collectives and report it to users.

Also we will add more unit tests to this script and make it modularized so that we can better maintain it.



[ghstack-poisoned]

fduwjj committed:
Update on "[RFC + WIC][Flight Recorder] Add more basic analysis to the script"


This is the first step to make sure we have a basic function of analyzer for FR in production.

- We want to use this script to find out abnormalities in collectives and report it to users.
- We also fixed some type errors.

- [Ongoing] Also we will add more unit tests to this script and make it modularized so that we can better maintain it.



[ghstack-poisoned]

fduwjj committed:
Update on "[RFC + WIC][Flight Recorder] Add more basic analysis to the script"


This is the first step to make sure we have a basic function of analyzer for FR in production.

- We want to use this script to find out abnormalities in collectives and report it to users.
- We also fixed some type errors.

- [Ongoing] Also we will add more unit tests to this script and make it modularized so that we can better maintain it.



[ghstack-poisoned]

c-p-i-o commented on tools/flight_recorder/fr_trace.py:
```diff
@@ -626,26 +719,82 @@ def build_collectives(
                 nccl_calls.extend(reversed(reversed_calls))
-
         else:
+            mismatch_count = 1
```
unused at the moment?

c-p-i-o reviewed (approved):
LGTM.

fduwjj committed:
Update on "[RFC + WIC][Flight Recorder] Add more basic analysis to the script"


This is the first step to make sure we have a basic function of analyzer for FR in production.

- We want to use this script to find out abnormalities in collectives and report it to users.
- We also fixed some type errors.

- [Ongoing] Also we will add more unit tests to this script and make it modularized so that we can better maintain it.



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[Flight Recorder] Add more basic analysis to the script"


This is the first step to make sure we have a basic function of analyzer for FR in production.

- We want to use this script to find out abnormalities in collectives and report it to users.
- We also fixed some type errors.

- [Ongoing] Also we will add more unit tests to this script and make it modularized so that we can better maintain it.



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented:
@pytorchbot rebase

fduwjj committed:
Update

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

fduwjj commented:
@pytorchbot merge -i

atalman commented:
@pytorchmergebot revert -c ifnoredsignal -m "New test: distributed/flight_recorder/test_fr_analysis is constantly failing"

atalman commented:
@pytorchmergebot revert -c ignoredsignal -m "New test: distributed/flight_recorder/test_fr_analysis is constantly failing"

fduwjj commented:
Interesting, the test passed for me locally.. Let me look into this.

fduwjj committed:
Update on "[Flight Recorder] Add more basic analysis to the script"


This is the first step to make sure we have a basic function of analyzer for FR in production.

- We want to use this script to find out abnormalities in collectives and report it to users.
- We also fixed some type errors.

- [Ongoing] Also we will add more unit tests to this script and make it modularized so that we can better maintain it.



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[Flight Recorder] Add more basic analysis to the script"


This is the first step to make sure we have a basic function of analyzer for FR in production.

- We want to use this script to find out abnormalities in collectives and report it to users.
- We also fixed some type errors.

- [Ongoing] Also we will add more unit tests to this script and make it modularized so that we can better maintain it.



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[Flight Recorder] Add more basic analysis to the script"


This is the first step to make sure we have a basic function of analyzer for FR in production.

- We want to use this script to find out abnormalities in collectives and report it to users.
- We also fixed some type errors.

- [Ongoing] Also we will add more unit tests to this script and make it modularized so that we can better maintain it.



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented on .ci/docker/requirements-ci.txt:
```diff
@@ -299,6 +299,9 @@ tensorboard==2.13.0
 #test that import: test_tensorboard
 
+#Pretty-print tabular data in Python, a library and a command-line utility.
+tabulate==0.9.0
```
We need this for CI to pass.

fduwjj replied:
Use a try-catch instead. Because that module will not be tested.

fduwjj committed:
Update on "[Flight Recorder] Add more basic analysis to the script"


This is the first step to make sure we have a basic function of analyzer for FR in production.

- We want to use this script to find out abnormalities in collectives and report it to users.
- We also fixed some type errors.

- [Ongoing] Also we will add more unit tests to this script and make it modularized so that we can better maintain it.



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[Flight Recorder] Add more basic analysis to the script"


This is the first step to make sure we have a basic function of analyzer for FR in production.

- We want to use this script to find out abnormalities in collectives and report it to users.
- We also fixed some type errors.

- [Ongoing] Also we will add more unit tests to this script and make it modularized so that we can better maintain it.



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented:
@pytorchbot rebase

fduwjj committed:
Update

[ghstack-poisoned]

fduwjj committed:
Update on "[Flight Recorder] Add more basic analysis to the script"


This is the first step to make sure we have a basic function of analyzer for FR in production.

- We want to use this script to find out abnormalities in collectives and report it to users.
- We also fixed some type errors.

- [Ongoing] Also we will add more unit tests to this script and make it modularized so that we can better maintain it.



cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

atalman reviewed (approved):
lgtm

fduwjj commented:
@pytorchbot merge -i

wconstab commented on tools/flight_recorder/fr_trace.py:
```diff
@@ -170,6 +187,50 @@ class Database(NamedTuple):
+    UNDECIDED = 5
+
+
+def check_size_even_expand(list1: List[Any], list2: List[Any], size: int) -> bool:
```
i think you're referring to 'broadcasting' by "expand" - maybe rename? (and maybe there is a torch helper existing for checking size including broadcasting? but no big deal..

fduwjj replied:
Oh I didn't see any existing helper function, yes, let me rename it.

wconstab commented on tools/flight_recorder/fr_trace.py:
```diff
@@ -170,6 +187,50 @@ class Database(NamedTuple):
+    """
+    Enum representing the possible states of matching for collective operations.
+
+    - FULLY_MATCHED: Indicates that all aspects of the collective operations match.
```
Nit: these strings seem useful for reporting error type to the user too.  should we attach them to the enum somehow or make a dict from enum to string to be able to print these out to the user at the appropriate time?

fduwjj replied:
I see. Yeah let's add it.

wconstab commented on tools/flight_recorder/fr_trace.py:
```diff
@@ -235,42 +297,79 @@ def __repr__(self) -> str:
+            if self.type in ["alltoall", "alltoall_base"]:
+                return MatchState.UNDECIDED
+            if self.type != "scatter" and self.input_sizes != other.input_sizes:
+                return MatchState.SIZE_OR_SYNTAX_MISMATCH
```
nit: feels like we could capture even more information in the MatchState object.

I wonder if we should actually construct an instance of MatchState and return that.  The instance could contain a 'mismatch string' or some lists of items A and items B where later on, someone analyzing the MatchState can see what the mismatch specifically was.

fduwjj replied:
Got it, sure. We can do that.

wconstab commented on tools/flight_recorder/fr_trace.py:
```diff
@@ -626,26 +729,81 @@ def build_collectives(
+                    f"Missing ranks are {expected_ranks - (candidate_ranks | found_ranks)}",
+                )
+            elif len(candidate_ranks) == 1:
+                # case two: alltoall or alltoall_base case.
```
for my education, why was alltoall a tricky case (or why do we mark it UNDECIDED) etc?

fduwjj replied:
Because you need to have wait all ranks to check input size and out size match. The input list and output list of alltoall only require the total numel of input and output to be the same.
